### ISSUE TITLE
About using a new dexhand

### ISSUE THREAD

**Issue by goodfor7** [ManipTrans/ManipTrans#35]
Hi!

Thanks to your help, I have successfully implemented the operation of training using my own collected motion capture data.

I am now in the process of applying this method to my own dexhand. In https://github.com/ManipTrans/ManipTrans/issues/15#issuecomment-2811735616 you  gave some advice. But I still have some questions.

The first is that the position of the wrist in the urdf of my hand is not at the origin, and in the mano2dexhand visualization the position of the wrist keeps changing as iter increases, at what point should I determine the Relative translation to align the mano with the dexhand.

Another question is whether the size of the dexhand is also related to the keypoint alignment in mano2dexhand, and if so, what to set to fix the alignment.

Thanks again for your patience in answering!

**Comment by KailinLi** [comment]
Hi — great to hear you have successfully adapted the pipeline to your new data!

* **Alignment**. Both `relative_rotation` and `relative_tsl` are fixed offsets between your dexterous hand and the MANO mesh; they are intrinsic properties and should not change as the MANO hand moves. You can fine-tune them as illustrated in issue [#33](https://github.com/ManipTrans/ManipTrans/issues/33#issuecomment-2982853928).

* **Size**. Because our method focuses more on fingertip tracking, it naturally mitigates most morphological differences. For a concrete example, please see our implementation for the Allegro Hand.

Hope this helps!

**Comment by goodfor7** [comment]
Thanks for the reply! I will try it later.

**Comment by goodfor7** [comment]
Your reply really helped me a lot!

I'm now at the step of getting imitators. In https://github.com/ManipTrans/ManipTrans/issues/3#issuecomment-2769172716 you mentioned using
`python main/rl/train.py dexhand=inspire task=DexHandImitator side=RH headless=true num_envs=5120 test=false randomStateInit=true dataIndices=[20aed@0,e9aab@2,925aa@1] learning_rate=2e-4 actionsMovingAverage=0.4 usePIDControl=False`

Can I use these datasets to train my own dexhand's imitator? But when I run `python main/dataset/mano2dexhand.py --data_idx e9aab@2 --dexhand MyHand --iter 7000` I find that the keypoints are difficult to align. Even if I reset self.relative_rotation (looks like e9aab@2 needs a different self.relative_rotation than my own motion capture data) , the situation remains unimproved. 

Does this mean I need to collect my own data to train the imitator？

And in https://github.com/ManipTrans/ManipTrans/issues/25#issuecomment-2846640695 you mentioned **_Including mirrored sequences, we trained on approximately 6–7 k demonstrations covering various task settings. On a single RTX 4090, the imitator converged after about three days._** Do I need 6-7k demonstrations to train my own dexhand as well? Is the operation to be performed by adding the dataset number to **_dataIndices=[20aed@0,e9aab@2,925aa@1]_**?

**Comment by KailinLi** [comment]
Hi! Two quick pointers that may help:
	1.	Key-point alignment – First, try training only on sequence e9aab@2 and deliberately over-fit to it. If everything is wired correctly, the imitator should track this single demo accurately.
	2.	Dataset size – Yes, we observed that ≈ 6–7 k demonstrations noticeably improve tracking quality. Once you have gathered enough sequences, list them all in the dataIndices=[ … ] argument, e.g. dataIndices=[20aed@0, \<your thousands of sequence IDs here\>…].

Hope that clarifies things!

**Comment by goodfor7** [comment]
Thank you for your reply. Still some questions remain:

1. When using data from different datasets, the rotation of the dexhand may be different, and does that have an effect on imitator training. But the setting of **_self.relative_rotation_** is global.

2. What is the approximate final reward for training the imitator? When I finished running
`python main/rl/train.py dexhand=myhand task=DexHandImitator side=RH headless=true num_envs=5120 test=false randomStateInit=true dataIndices=[20aed@0,e9aab@2,925aa@1] learning_rate=2e-4 actionsMovingAverage=0.4 usePIDControl=False`
on 4090 for 3 days, the reward value is about 22. Later when I used this imitator to train dexhand for the 20aed@0 operation, after 150 epochs, the reward value was only 4.4 (initially 4.0). Obviously 20aed@0 is also included in the training of the imitator, so why is this happening?

3. Is the preprocessing step
`python main/dataset/mano2dexhand.py --data_idx 20aed@0 --side right --dexhand inspire --headless --iter 7000`
still needed before training imitator

**Comment by KailinLi** [comment]
Hello!
1.	`relative_rotation` represents the fixed offset between the dexterous-hand model and the MANO hand. As long as your dataset follows the MANO convention, the same relative_rotation should apply.
2.	A final reward of 4.4 is indeed unusual. Please double-check the preprocessing output—especially the initial states—to see whether severe hand-object penetrations or other issues occur right at the start of training.
3.	For the imitator you do not need to run the mano2dexhand.py preprocessing step. The imitator resets each episode with random joint configurations for robustness, so it never uses preprocessing artifacts such as the `opt_*` values (see the `_reset_default` function).

Hope this helps!

**Comment by goodfor7** [comment]
Hi! Thanks for your advice!

These days, I tried training imitator on e9aab@2 using shadowhand and was eventually able to perform the operation properly. I guess there might be a problem with the structure or correspondence of my dexhand.

Also, based on the datasets you mentioned in your paper for training imitator, I tried to migrate the method to the GRAB datasets. The reward for training tops out at about 2200 after 1500 epochs, but I'm having problems with testing as in the video (after disabling **_failed_execute_**, the hand keeps returning to the initial stance if it's not disabled).

Since the origin is not the table in the npz file, I directly subtracted the **_transl_** of the rhand and object from the **_transl_** of the table and did not process the **_global_orient_** (is this the cause of the problem?). I've kept the format of the final desired file consistent with **_102_sv_dict.npy_** and continues to use **_grab_dataset_dexhand.py_** . I'm using raw 120fps and when I set **_skip=2_** , then the hand doesn't touch the object at all. 


        "obj_trajectory": torch.tensor(
                np.stack(obj_pose[:: self.skip].cpu()), device=self.device, dtype=torch.float
            ),
        "wrist_pos": wrist_pos[:: self.skip],
        "wrist_rot": wrist_rot[:: self.skip],
        "mano_joints": {key: value[:: self.skip] for key, value in mano_joints.items()},

https://github.com/user-attachments/assets/47bc716c-83f0-4825-8b34-4dee2624c555

**Comment by KailinLi** [comment]
Hi — great to see your progress!
	•	The failed_execute flag is typically triggered by a large tracking error during rollout, which can happen if the policy hasn’t seen enough diverse data during training.
	•	For imitation learning, global_orient is important. Please align the global orientation (and translation) of the hand, object, and table into the same world frame; subtracting translations alone without handling orientation can introduce drift.
	•	Regarding frame rate: if downsampling from 120 fps to 60 fps (skip=2) causes the hand to miss contact, this sequence of hand movement may simply be too fast to track?

Hope this helps.

**Comment by goodfor7** [comment]
Hi!

Based on my previous attempt, I tried cutting the first 150 frames of the sequence and setting skip=2. The remaining actions were executed normally without triggering failed_execute, but the last part of the action still seemed to trigger failed_execute. So I guessed that it might be due to the action sequence being too long or the hand being too far away from the table and the object? The arm was flat at the beginning and end, as the picture shows.

<img width="1110" height="958" alt="Image" src="https://github.com/user-attachments/assets/2ec90d05-08dd-4f41-84d4-6dffabcd5153" />

**Comment by KailinLi** [comment]
Hi! Thanks for the detailed report and the image.

Your hypothesis makes sense. ManipTrans’s training/hyperparameters were tailored for small spatial displacements and fine-grained manipulation, so **long-reach, high-speed pre-grasp approach trajectories** fall outside its optimal operating regime and can trigger failed_execute near the end.

The codebase can, in principle, track larger excursions, but the reward shaping and certain motion/control parameters (e.g., velocity/acceleration limits, action smoothing, tracking gains) likely need to be retuned specifically for fast approaches.

As a practical note, in our real-world experiments, many robotic arms cannot execute human-arm-like, long-range, high-speed, high-precision moves; such performance typically requires specialized industrial manipulators.

**Comment by goodfor7** [comment]
Thank you for your detailed explanation. I think I can understand what you mean.

I also tried applying this method to the dexycb datasets. However, I encountered an issue that was difficult to resolve. As shown in the figure, in mano2dexhand, the relative positions of the hand and the object are roughly correct, but there is an issue during the training phase. The failed_execute trigger was activated, and when I disabled it, I found that the hand's trajectory (green) was normal, but the object was not being grasped by the hand. When I visualized the training, I noticed that the object flies upward as soon as it is touched by the hand. Is this a normal occurrence?

In the original dexycb, the trajectories of the hand and object are tilted relative to the coordinate axes. I used the AprilTag provided by the author to handle them. Regardless of whether the data has been processed or not, there seems to be some gap between the hand's keypoints and the object. Is this the issue? Or is there a problem with the urdf file I wrote?

How should I resolve this issue?

```
<?xml version="1.0"?>
<robot name="scan">
  <material name="obj_color">
      <color rgba="1.0 0.423529411765 0.0392156862745 1.0"/>
  </material>
  <link name="base">
    <visual>
      <origin xyz="0.0 0.0 0.0"/>
      <geometry>
        <mesh filename="textured.obj" scale="1 1 1"/>
      </geometry>
      <material name="obj_color"/>
    </visual>
    <collision>
      <origin xyz="0.0 0.0 0.0"/>
      <geometry>
        <mesh filename="textured.obj" scale="1 1 1"/>
      </geometry>
    </collision>
  </link>
</robot>
```

<img width="1122" height="1016" alt="Image" src="https://github.com/user-attachments/assets/72122fce-dc44-46d3-9ed0-58fe728600a7" />

<img width="1347" height="866" alt="Image" src="https://github.com/user-attachments/assets/78c09a9f-b49c-4c83-8415-ae32bb649994" />

**Comment by goodfor7** [comment]
Thank you for your detailed explanation and patience.

**Comment by yys-abu** [comment]
> Hi! Thanks for your advice!
> 
> These days, I tried training imitator on e9aab@2 using shadowhand and was eventually able to perform the operation properly. I guess there might be a problem with the structure or correspondence of my dexhand.
> 
> Also, based on the datasets you mentioned in your paper for training imitator, I tried to migrate the method to the GRAB datasets. The reward for training tops out at about 2200 after 1500 epochs, but I'm having problems with testing as in the video (after disabling **_failed_execute_**, the hand keeps returning to the initial stance if it's not disabled).
> 
> Since the origin is not the table in the npz file, I directly subtracted the **_transl_** of the rhand and object from the **_transl_** of the table and did not process the **_global_orient_** (is this the cause of the problem?). I've kept the format of the final desired file consistent with **_102_sv_dict.npy_** and continues to use **_grab_dataset_dexhand.py_** . I'm using raw 120fps and when I set **_skip=2_** , then the hand doesn't touch the object at all.
> 
> ```
>     "obj_trajectory": torch.tensor(
>             np.stack(obj_pose[:: self.skip].cpu()), device=self.device, dtype=torch.float
>         ),
>     "wrist_pos": wrist_pos[:: self.skip],
>     "wrist_rot": wrist_rot[:: self.skip],
>     "mano_joints": {key: value[:: self.skip] for key, value in mano_joints.items()},
> ```
> 
>  redpandacompress_DESKTOP-U01TI6C.2025-07-25.16-56-50.mp4

May I ask whether, once training is complete, it is only possible to visualise the results, and not save the converted data?

**Comment by goodfor7** [comment]
> > Hi! Thanks for your advice!
> > These days, I tried training imitator on e9aab@2 using shadowhand and was eventually able to perform the operation properly. I guess there might be a problem with the structure or correspondence of my dexhand.
> > Also, based on the datasets you mentioned in your paper for training imitator, I tried to migrate the method to the GRAB datasets. The reward for training tops out at about 2200 after 1500 epochs, but I'm having problems with testing as in the video (after disabling **_failed_execute_**, the hand keeps returning to the initial stance if it's not disabled).
> > Since the origin is not the table in the npz file, I directly subtracted the **_transl_** of the rhand and object from the **_transl_** of the table and did not process the **_global_orient_** (is this the cause of the problem?). I've kept the format of the final desired file consistent with **_102_sv_dict.npy_** and continues to use **_grab_dataset_dexhand.py_** . I'm using raw 120fps and when I set **_skip=2_** , then the hand doesn't touch the object at all.
> > ```
> >     "obj_trajectory": torch.tensor(
> >             np.stack(obj_pose[:: self.skip].cpu()), device=self.device, dtype=torch.float
> >         ),
> >     "wrist_pos": wrist_pos[:: self.skip],
> >     "wrist_rot": wrist_rot[:: self.skip],
> >     "mano_joints": {key: value[:: self.skip] for key, value in mano_joints.items()},
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > redpandacompress_DESKTOP-U01TI6C.2025-07-25.16-56-50.mp4
> 
> May I ask whether, once training is complete, it is only possible to visualise the results, and not save the converted data?

I recall the author provided a method for saving data in other questions.

**Comment by lililifn** [comment]
> 你好！
> 
> 多亏了你的帮助，我成功地利用自己收集的动作捕捉数据实现了训练作。
> 
> 我现在正在把这种方法应用到我自己的敏捷手上。在[#15（评论）](https://github.com/ManipTrans/ManipTrans/issues/15#issuecomment-2811735616)中你给了一些建议。但我仍然有一些问题。
> 
> 第一个是手腕在手的右侧手部位置不在原点，在手握手的视觉化中，手腕的位置随着手腕的增加不断变化，我应该在什么时候确定相对平移来使手部与直手对齐。
> 
> 另一个问题是，dexhand的大小是否也与mano2dexhand的关键点对齐有关，如果有，应该设置什么来固定对齐。
> 
> 再次感谢你的耐心回答！

Hello！I also plan to port this method to a new dexterous hand, 
but I am unsure how to obtain a dataset for the new dexterous hand. 
Could you please share some projects related to acquiring dexterous hand grasping datasets for reference? I haven't found any relevant open-source files so far, and I would greatly appreciate your help.

**Comment by goodfor7** [comment]
> > 你好！
> > 多亏了你的帮助，我成功地利用自己收集的动作捕捉数据实现了训练作。
> > 我现在正在把这种方法应用到我自己的敏捷手上。在[#15（评论）](https://github.com/ManipTrans/ManipTrans/issues/15#issuecomment-2811735616)中你给了一些建议。但我仍然有一些问题。
> > 第一个是手腕在手的右侧手部位置不在原点，在手握手的视觉化中，手腕的位置随着手腕的增加不断变化，我应该在什么时候确定相对平移来使手部与直手对齐。
> > 另一个问题是，dexhand的大小是否也与mano2dexhand的关键点对齐有关，如果有，应该设置什么来固定对齐。
> > 再次感谢你的耐心回答！
> 
> Hello！I also plan to port this method to a new dexterous hand, but I am unsure how to obtain a dataset for the new dexterous hand. Could you please share some projects related to acquiring dexterous hand grasping datasets for reference? I haven't found any relevant open-source files so far, and I would greatly appreciate your help.

If you're referring to manual data, I use motion capture equipment.If it is dexterous hand data, then it is the final output of this method.
